libfwupdplugin/fu-smbios.c

```diff
@@ -341,6 +341,147 @@ fu_smbios_setup_from_file (FuSmbios *self, const gchar *filename, GError **error
+	return TRUE;
+}
+
+/*
+ * The mapping from SMBIOS field to sysfs name can be found by mapping
+ * the field to a kernel property name in dmi_decode()
+ * (drivers/firmware/dmi_scan.c), then the property name to sysfs entry
+ * in dmi_id_init_attr_table() (drivers/firmware/dmi-id.c). This table
+ * lists each attribute exposed in /sys/class/dmi when CONFIG_DMIID is
+ * enabled, mapping to the SMBIOS field and a function that can convert
+ * the textual version of the field back into the raw SMBIOS table
+ * representation.
+ */
+#define SYSFS_DMI_FIELD(_name, _type, _offset, kind) \
+	{ .name = _name, .type = _type, .offset = _offset, .encode = fu_smbios_encode_ ## kind ## _from_kernel }
+const struct kernel_dmi_field {
+	const gchar *name;
+	gboolean (*encode)(FuSmbios *, const gchar *, guint8, guint8, GError **);
+	guint8 type;
+	guint8 offset;
+} KERNEL_DMI_FIELDS[] = {
+	SYSFS_DMI_FIELD("bios_vendor", 0, 4, string),
+	SYSFS_DMI_FIELD("bios_version", 0, 5, string),
+	SYSFS_DMI_FIELD("bios_date", 0, 8, string),
+	SYSFS_DMI_FIELD("sys_vendor", 1, 4, string),
+	SYSFS_DMI_FIELD("product_name", 1, 5, string),
+	SYSFS_DMI_FIELD("product_version", 1, 6, string),
+	SYSFS_DMI_FIELD("product_serial", 1, 7, string),
+	/* SYSFS_DMI_FIELD("product_uuid", 1, 8, uuid) */
+	SYSFS_DMI_FIELD("product_family", 1, 26, string),
+	SYSFS_DMI_FIELD("product_sku", 1, 25, string),
+	SYSFS_DMI_FIELD("board_vendor", 2, 4, string),
+	SYSFS_DMI_FIELD("board_name", 2, 5, string),
+	SYSFS_DMI_FIELD("board_version", 2, 6, string),
+	SYSFS_DMI_FIELD("board_serial", 2, 7, string),
+	SYSFS_DMI_FIELD("board_asset_tag", 2, 8, string),
+	SYSFS_DMI_FIELD("chassis_vendor", 3, 4, string),
+	SYSFS_DMI_FIELD("chassis_type", 3, 5, byte),
+	SYSFS_DMI_FIELD("chassis_version", 3, 6, string),
+	SYSFS_DMI_FIELD("chassis_serial", 3, 7, string),
+	SYSFS_DMI_FIELD("chassis_asset_tag", 3, 8, string),
+};
+
+/**
+ * fu_smbios_setup_from_kernel:
+ * @self: a #FuSmbios
+ * @filename: a directory path
+ * @error: (nullable): optional return location for an error
+ *
+ * Reads SMBIOS value from DMI values provided by the kernel, such as in
+ * /sys/class/dmi on Linux.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.6.2
+ **/
+static gboolean
+fu_smbios_setup_from_kernel (FuSmbios *self, const gchar *class_path, GError **error)
+{
+	gboolean any_success = FALSE;
+
+	/* add fake structures */
+	for (guint i = 0; i < FU_SMBIOS_STRUCTURE_TYPE_LAST; i++) {
+		FuSmbiosItem *item = g_new0 (FuSmbiosItem, 1);
+		item->type = i;
+		item->buf = g_byte_array_new ();
+		item->strings = g_ptr_array_new_with_free_func (g_free);
+		g_ptr_array_add (self->items, item);
+	}
+
+	/* parse every known field from the corresponding file */
+	for (gsize i = 0; i < (sizeof (KERNEL_DMI_FIELDS) / sizeof (*KERNEL_DMI_FIELDS)); i++) {
+		const struct kernel_dmi_field *field = &KERNEL_DMI_FIELDS[i];
+		g_autofree gchar *path = g_build_filename (class_path, field->name, NULL);
+		GError *local_error = NULL;
```

Comment:
this should be a `g_autoptr`, and then no need to clear the error each time in the error scenarios.

Reply:
Done.